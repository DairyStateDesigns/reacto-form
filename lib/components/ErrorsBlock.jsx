import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEmpty from 'lodash.isempty';

import customPropTypes from '../shared/propTypes';

class ErrorsBlock extends Component {
  static isComposableFormErrors = true;

  static propTypes = {
    // Additional classes to add to the block div
    className: PropTypes.string,
    errors: customPropTypes.errors,
    // Styles object for the individual error messages in the block
    errorStyle: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    names: PropTypes.arrayOf(PropTypes.string),
    // Styles object for the block div
    style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  };

  static defaultProps = {
    className: null,
    errors: undefined,
    errorStyle: {},
    names: undefined,
    style: {},
  };

  render() {
    const { errorStyle, errors, className: classNameProp, style } = this.props;
    const className = `reacto-form-errors ${classNameProp || ''}`.trim();

    if (isEmpty(errors)) return null;

    // https://reactjs.org/docs/lists-and-keys.html
    // "When you don’t have stable IDs for rendered items, you may use the item index as a key as a last resort.
    // We don’t recommend using indexes for keys if the items can reorder, as that would be slow."
    //
    // There will rarely be more than a few errors for a field, and forcing unique ID
    // generation for them would be arbitrary and unnecessary. So we'll use the index.
    return (
      <div className={className} style={style}>
        {errors.map((error, index) => {
          return (<div key={index} data-name={error.name} style={errorStyle}>{error.message}</div>); // eslint-disable-line react/no-array-index-key
        })}
      </div>
    );
  }
}

export default ErrorsBlock;
