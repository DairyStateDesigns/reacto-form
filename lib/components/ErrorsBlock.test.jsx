import React from 'react';
import renderer from 'react-test-renderer';
import ErrorsBlock from './ErrorsBlock';

test('has isComposableFormErrors property set to true', () => {
  expect(ErrorsBlock.isComposableFormErrors).toBe(true);
});

test('renders all error messages', () => {
  const errors = [
    { name: 'a', message: 'Message One' },
    { name: 'b', message: 'Message Two' },
  ];

  const component = renderer.create(
    <ErrorsBlock errors={errors} />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders nothing when there are no errors', () => {
  const component = renderer.create(
    <ErrorsBlock />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
