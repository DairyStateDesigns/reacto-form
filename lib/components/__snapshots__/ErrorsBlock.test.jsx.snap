// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all error messages 1`] = `
<div
  className={undefined}
  style={Object {}}
>
  <div
    className={undefined}
    data-name="a"
    style={Object {}}
  >
    Message One
  </div>
  <div
    className={undefined}
    data-name="b"
    style={Object {}}
  >
    Message Two
  </div>
</div>
`;

exports[`renders nothing when there are no errors 1`] = `null`;
